mesonpy/__init__.py

```diff
@@ -1138,7 +1138,12 @@ def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
         try:
             return func(*args, **kwargs)
         except Error as exc:
-            print('{red}meson-python: error:{reset} {msg}'.format(msg=str(exc), **_STYLES))
+            print((
+                '{red}meson-python: error:{reset}\n'
+                '{red}>{reset}\n' +
+                textwrap.indent(str(exc).strip(), prefix='{red}>{reset}  ') + '\n' +
+                '{red}>{reset}\n'
+            ).format(**_STYLES))
             raise SystemExit(1) from exc
```

Comment:
This makes all errors printed by meson look like:
```
meson-python: error:
> 
>  Error message
>
```
turning one line error messages into four lines error messages wasting three lines for unnecessary decoration and makes meson-python errors harder to grep for. I don't expect that meson-python will be emitting many very long error messages, thus this format does not seem optimal.

I would prefer to use the same prefix for all lines, without decoration:
```
meson-python: error: Error message
```
with multi-line messages (which I guess are emitted by newer versions of pyproject-metadata, as meson-python itself does not emit any multi-line error messages), formatted like this:
```
meson-python: error: Error message first line
meson-python: error: Error message second line
meson-python: error: Even more lines...
```
This is more compact, immediate to grep for, and works equally well for reporting multiple unrelated errors and single error conditions with details spanning multiple lines.

Reply:
> I don't expect that meson-python will be emitting many very long error messages, thus this format does not seem optimal.

We will once `pyproject-metadata` is able to report all validation errors at once (see https://github.com/FFY00/python-pyproject-metadata/issues/45).

> I would prefer to use the same prefix for all lines, without decoration:
> 
> ```
> meson-python: error: Error message
> ```
> 
> with multi-line messages (which I guess are emitted by newer versions of pyproject-metadata, as meson-python itself does not emit any multi-line error messages), formatted like this:
> 
> ```
> meson-python: error: Error message first line
> meson-python: error: Error message second line
> meson-python: error: Even more lines
> ```

This looks like multiple errors, and looks weird on big messages.

> This is more compact, immediate to grep for,

But the error is the last thing we will print after exiting, and the `meson-python: error:` prefix isn't used anywhere here... I don't understand why people would be grepping, or at least why that'd be a use-case we care about.

> and works equally well for reporting multiple unrelated errors and single error conditions with details spanning multiple lines.

But that's not something we do, or need, and I don't think we will, I can't really think of a reason why we'd have multiple different kinds of errors.

---

I am okay with keeping the single line format on single line errors, and only opting for this one when we have multi-line error descriptions. What do you think?

Reply:
So we are making the logging of an error message with a single line of text looks ridiculous to optimize for an hypothetical future in which one of our dependencies will report multiple errors?

> I am okay with keeping the single line format on single line errors, and only opting for this one when we have multi-line error descriptions. What do you think?

Let's wait till there is the need to report multi-line errors to add this. At the moment it is not even clear in which form multiple validation errors will be reported by `pyproject-metadata`.

Reply:
> Let's wait till there is the need to report multi-line errors to add this.

If the logic is going to if/else how to print it based on the presence of newlines in the message, then it should be fine to optimistically add support for multi-line errors without changing single-line errors. Even if the code is then never actually used, it's just "building infrastructure for the future". :D

Reply:
> So we are making the logging of an error message with a single line of text looks ridiculous to optimize for an hypothetical future in which one of our dependencies will report multiple errors?

> If the logic is going to if/else how to print it based on the presence of newlines in the message, then it should be fine to optimistically add support for multi-line errors without changing single-line errors. Even if the code is then never actually used, it's just "building infrastructure for the future". :D

`_pyproject_hook` currently only triggers on our errors, which we do control, but it should actually be triggering on *all* expected errors. As such, we should not assume the errors are single line.

> Let's wait till there is the need to report multi-line errors to add this.

I really don't understand the issue with keeping the format for single line errors and just handling multi-line errors better.

> At the moment it is not even clear in which form multiple validation errors will be reported by `pyproject-metadata`.

I don't think this is very relevant, but since I am the one implementing it, I can tell you it will raise an error something like this:

> Found several invalid keys in pyproject.toml:
>   - project.name: expecting a string, but got "None"
>   - project.license: the license file was not found
>   -  ...

It may be an `ExceptionGroup` with a custom repr, or a normal exception (with the same repr).

Reply:
So I think this is a very minor point in the end. I don't have a clear preference, but I suggest simply removing the two lines containing `'{red}>{reset}\n'` to avoid the discussion and the extra two lines in the output that @dnicolodi does not like. 

Right now the error for a single config mistake by the user is:
```
meson-python: error:
>
>  Field `project.name` missing
>
```
It could be changed back to:
```
meson-python: error: Field `project.name` missing
```

It really doesn't matter too much either way. The gain of this PR is the exception chaining, so the huge traceback gets hidden. I'd like to get that merged.